Index: xserver-xorg-input-synaptics/tools/synclient.c
===================================================================
--- xserver-xorg-input-synaptics.orig/tools/synclient.c	2011-07-05 11:54:31.731347670 +0300
+++ xserver-xorg-input-synaptics/tools/synclient.c	2011-07-05 11:55:06.211347676 +0300
@@ -391,9 +391,7 @@
 unwind:
     XFree(properties);
     XFreeDeviceList(info);
-    if (!dev)
-        fprintf(stderr, "Unable to find a synaptics device.\n");
-    else if (error && dev)
+    if (error && dev)
     {
 	XCloseDevice(dpy, dev);
 	dev = NULL;
Index: xserver-xorg-input-synaptics/tools/syndaemon.c
===================================================================
--- xserver-xorg-input-synaptics.orig/tools/syndaemon.c	2011-07-05 11:54:31.751347670 +0300
+++ xserver-xorg-input-synaptics/tools/syndaemon.c	2011-07-05 11:55:06.211347676 +0300
@@ -512,9 +512,7 @@
 unwind:
     XFree(properties);
     XFreeDeviceList(info);
-    if (!dev)
-	fprintf(stderr, "Unable to find a synaptics device.\n");
-    else if (error && dev)
+    if (error && dev)
     {
 	XCloseDevice(dpy, dev);
 	dev = NULL;
